tests/test_datastores.py

```diff
@@ -0,0 +1,368 @@
+    """Check that the `datastore.get_xy` method is implemented."""
+    datastore = init_datastore_example(datastore_name)
+
+    if not isinstance(datastore, BaseRegularGridDatastore):
+        pytest.skip("Datastore does not implement `BaseCartesianDatastore`")
+
+    for category in ["state", "forcing", "static"]:
+        xy_stacked = datastore.get_xy(category=category, stacked=True)
+        xy_unstacked = datastore.get_xy(category=category, stacked=False)
+
+        assert isinstance(xy_stacked, np.ndarray)
+        assert isinstance(xy_unstacked, np.ndarray)
+
+        nx, ny = datastore.grid_shape_state.x, datastore.grid_shape_state.y
+
+        # for stacked=True, the shape should be (n_grid_points, 2)
+        assert xy_stacked.ndim == 2
+        assert xy_stacked.shape[0] == nx * ny
+        assert xy_stacked.shape[1] == 2
+
+        # for stacked=False, the shape should be (nx, ny, 2)
+        assert xy_unstacked.ndim == 3
+        assert xy_unstacked.shape[0] == nx
+        assert xy_unstacked.shape[1] == ny
+        assert xy_unstacked.shape[2] == 2
+
+
+@pytest.mark.parametrize("datastore_name", DATASTORES.keys())
+def test_get_projection(datastore_name):
+    """Check that the `datastore.coords_projection` property is implemented."""
+    datastore = init_datastore_example(datastore_name)
+
+    if not isinstance(datastore, BaseRegularGridDatastore):
+        pytest.skip("Datastore does not implement `BaseCartesianDatastore`")
+
+    assert isinstance(datastore.coords_projection, ccrs.Projection)
+
+
+@pytest.mark.parametrize("datastore_name", DATASTORES.keys())
+def get_grid_shape_state(datastore_name):
+    """Check that the `datastore.grid_shape_state` property is implemented."""
+    datastore = init_datastore_example(datastore_name)
+
+    if not isinstance(datastore, BaseRegularGridDatastore):
+        pytest.skip("Datastore does not implement `BaseCartesianDatastore`")
+
+    grid_shape = datastore.grid_shape_state
+    assert isinstance(grid_shape, tuple)
+    assert len(grid_shape) == 2
+    assert all(isinstance(e, int) for e in grid_shape)
+    assert all(e > 0 for e in grid_shape)
+
+
+@pytest.mark.parametrize("datastore_name", DATASTORES.keys())
+def test_stacking_grid_coords(datastore_name):
+    """Check that the `datastore.stack_grid_coords` method is implemented."""
+    datastore = init_datastore_example(datastore_name)
+
+    if not isinstance(datastore, BaseRegularGridDatastore):
+        pytest.skip("Datastore does not implement `BaseCartesianDatastore`")
+
+    da_static = datastore.get_dataarray("static", split=None)
+
+    da_static_unstacked = datastore.unstack_grid_coords(da_static).load()
+    da_static_test = datastore.stack_grid_coords(da_static_unstacked)
+
+    # XXX: for the moment unstacking doesn't guarantee the order of the
+    # dimensions maybe we should enforce this?
+    da_static_test = da_static_test.transpose(*da_static.dims)
+
+    xr.testing.assert_equal(da_static, da_static_test)
```

Comment:
Ok, yes this is how I feel too. I will try and find out what is going wrong and then ensure that the dimension order is the same so that `datastore.stack_grid_coords(...)` returns the same dimension order always.

Reply:
I've tried to resolve this in https://github.com/mllam/neural-lam/pull/66/commits/b7a10efc582db323c6dbc9c178534a1432a9cfb0, what do you think? We haven't really discussed whether we should enforce a dimension order in what is returned from at datastore in general, maybe that would be a good idea to do. It doesn't matter if we do dimension-based operations (for example selection), and we could only really say in the docstring of the baseclass that people _should_ do this, we can't enforce it unless we add a check in `WeatherDataset` or add a test. I don't feel strongly that this is needed, but maybe I am missing something... (sorry about the force pushed, I had too much in the commit and it would have been hard for you to see the change)

Reply:
I think it's fine to just enforce the ordering returned here. No need to add an additional test, as the stack/unstack test should catch any inconsistencies. I would for consistency however prefer to place the (grid_index, feature_dim) last rather than first, as that is how all tensors are ordered and how we order them in build_item_dataarrays in WeatherDataset. Otherwise I think this change looks good!

Reply:
Ok, quick update on this: I think this is now resolved. I realised that to *ensure* the dimensions are in a specific order we either had to a) call `BaseDatastore.get_dataarray(...)` and get the order order of dimensions for a specific `category` of data and then use that order (this could be cached of course or b) implement some way of defining the dimension order and then apply that when returning at dataset from `BaseDatastore.get_dataarray(...)`. This is necessary because the order of dimensions depend on how the content of the zarr dataset is stored, which although we could just assume is always the same, I thought it safer not to.

So, in https://github.com/mllam/neural-lam/pull/66/commits/1121d9f12aeb34ab5e1d32c87d008201c9ec0754 I have added `BaseDatatstore.expected_dim_order(category=...)` which is now used both by `BaseDatastore.get_dataarray(category=...)` and `BaseRegularGridDatastore.stack_grid_coords(...)` to ensure that both return a `xr.DataArray` with the same order.

I later realised that I hadn't implemented this correctly and hadn't identified all the implications of adding this functionality (specifically in the MEPS datastore I previously concatenated along a dimension called `feature` after stacking in space, but stacking in space needs to be done with the feature dimension correctly named `{category}_feature`. I fixed this in https://github.com/mllam/neural-lam/pull/66/commits/d355ef5030c80faa50e4f440cddee052597dd53b)

Reply:
Hope this implementation makes sense!

Reply:
I think that implementation with `expected_dim_order` makes sense. Two things though:

1. Will it now be unnecessary to do the transposes in the dataset class (e.g. https://github.com/leifdenby/neural-lam/blob/0f24924aa45550f05f414bccf64139ba827dcb58/neural_lam/weather_dataset.py#L378), since the datastore will return consistent dimension orderings already?
2. If I am not misreading the code this still places  (grid_index, feature_dim) first instead of last. I think those dims should be placed last, in particular to be consistent with what is returned from the dataset class.

Reply:
> Will it now be unnecessary to do the transposes in the dataset class (e.g. https://github.com/leifdenby/neural-lam/blob/0f24924aa45550f05f414bccf64139ba827dcb58/neural_lam/weather_dataset.py#L378), since the datastore will return consistent dimension orderings already?

Good point. I think instead it would be better to simply assert the order in `WeatherDataset.__getitem__` since we are relying the user to remember to do the transpose before returning the `xr.DataArray`. I will change this.

> If I am not misreading the code this still places (grid_index, feature_dim) first instead of last. I think those dims should be placed last, in particular to be consistent with what is returned from the dataset class.

Ok good, yes that is fine with me, good idea. I will do that

Reply:
Looks great! Resolving this.